### PR TITLE
IDE-2081 add a "launch new workspace" action

### PR DESCRIPTION
**Pull request by AndyWu2015**

### CODE REVIEW

**Comment by gamerson** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

File: common/plugins/com.liferay.ide.ui/plugin.xml [hunk 1]
```diff
@@ -182,8 +182,8 @@
             id="com.liferay.ide.ui.workspace.launchWorkspace"
             name="Launch Workspace">
             <commandParameter
-                  id="com.liferay.ide.ui.workspace.launchWorkspace.workspaceLocation"
-                  name="Workspace Location"
+                  id="workspaceLocation"
```
Comment (gamerson) [inline common/plugins/com.liferay.ide.ui/plugin.xml]: Why did you need to update this?  I'm not opposed to these values but just wondering why you needed to change them?

File: common/plugins/com.liferay.ide.ui/src/com/liferay/ide/ui/dialog/ChooseWorkspaceWithPreferenceDialog.java [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+import org.eclipse.ui.internal.ide.IDEWorkbenchPlugin;
+import org.eclipse.ui.internal.ide.IIDEHelpContextIds;
+import org.eclipse.ui.internal.ide.StatusUtil;
+import org.eclipse.ui.preferences.SettingsTransfer;
+
+/**
+ * @author Andy Wu
+ */
+@SuppressWarnings( "restriction" )
+public class ChooseWorkspaceWithPreferenceDialog extends ChooseWorkspaceDialog
```
Comment (gamerson) [inline common/plugins/com.liferay.ide.ui/src/com/liferay/ide/ui/dialog/ChooseWorkspaceWithPreferenceDialog.java]: Could you extend ChooseWorkspaceWithSettingsDialog instead and reduce the amount of code that was copied from that class?

File: common/plugins/com.liferay.ide.ui/src/com/liferay/ide/ui/workspace/LaunchLiferay7WorkspaceAction.java [hunk 3]
```diff
@@ -14,8 +35,25 @@ public LaunchLiferay7WorkspaceAction()
...
+
     @Override
     public void run()
     {
+        LaunchWorkspaceHandler hanlder = new LaunchWorkspaceHandler();
+
+        Map<String, String> map = new HashMap<String, String>();
+
+        // TODO which workspace shall we point to ?
+        map.put( LaunchWorkspaceHandler.PARAM_WORKSPACE_LOCATION, "D:/test/forlaunch" );
```
Comment (gamerson) [inline common/plugins/com.liferay.ide.ui/src/com/liferay/ide/ui/workspace/LaunchLiferay7WorkspaceAction.java]: Can you put this value as anything?  Does it really matter?  Can you just use the string "workspace"?

File: common/plugins/com.liferay.ide.ui/src/com/liferay/ide/ui/workspace/LaunchWorkspaceMenu.java [hunk 4]
```diff
@@ -29,9 +44,9 @@ public void createContributionItems( IServiceLocator serviceLocator, IContributi
     {
         final MenuManager menu = new MenuManager( "Launch Workspace" );
 
-        menu.add( new LaunchLiferay7WorkspaceAction() );
+        //menu.add( new LaunchLiferay7WorkspaceAction() );
```
Comment (gamerson) [inline common/plugins/com.liferay.ide.ui/src/com/liferay/ide/ui/workspace/LaunchWorkspaceMenu.java]: We can remove this class if we aren't going to use it anymore.

File: common/plugins/com.liferay.ide.ui/src/com/liferay/ide/ui/workspace/WorkbenchPreferencesTransfer.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        super();
+    }
+
+    @Override
+    public IStatus transferSettings( IPath newWorkspaceRoot )
+    {
+        IPath currentWorkspace = Platform.getLocation();
+
+        File srcDir = new File( currentWorkspace.toFile(), ".metadata/.plugins/org.eclipse.core.runtime/.settings" );
+        File destDir = new File( newWorkspaceRoot.toFile(), ".metadata/.plugins/org.eclipse.core.runtime/.settings" );
```
Comment (gamerson) [inline common/plugins/com.liferay.ide.ui/src/com/liferay/ide/ui/workspace/WorkbenchPreferencesTransfer.java]: Are you sure this is safe to do?  Is this exactly the same as "export" all preferences and "import" those settings?  Is this setting to copy the settings like this?

**Comment by gamerson** [comment]
Pull request submitted to https://github.com/AndyWu2015/liferay-ide/pull/8. [See changes here.](https://github.com/gamerson/liferay-ide/compare/130fb66~1...130fb66) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

File: common/plugins/com.liferay.ide.ui/plugin.xml (hunk 1 above)
Comment (AndyWu2015) [inline common/plugins/com.liferay.ide.ui/plugin.xml]: it is a bug  , the name should be the same as LaunchWorkspaceHandler.PARAM_WORKSPACE_LOCATION , otherwise the history workspace menu contribution will not show out

Comment (AndyWu2015) [inline common/plugins/com.liferay.ide.ui/plugin.xml]: it is a bug  , the name should be the same as LaunchWorkspaceHandler.PARAM_WORKSPACE_LOCATION , otherwise the history workspace menu contribution will not show out

File: common/plugins/com.liferay.ide.ui/src/com/liferay/ide/ui/workspace/WorkbenchPreferencesTransfer.java (hunk 5 above)
Comment (AndyWu2015) [inline common/plugins/com.liferay.ide.ui/src/com/liferay/ide/ui/workspace/WorkbenchPreferencesTransfer.java]: I tested it someway and worked well but not very sure , I read the http://www.vogella.com/tutorials/EclipsePreferences/article.html 
and think this is almost the right way to do .

File: common/plugins/com.liferay.ide.ui/plugin.xml (hunk 1 above)
Comment (AndyWu2015) [inline common/plugins/com.liferay.ide.ui/plugin.xml]: I read the source code dealing with the menu issues and make sure it it necessary  to do this

File: common/plugins/com.liferay.ide.ui/src/com/liferay/ide/ui/workspace/WorkbenchPreferencesTransfer.java (hunk 5 above)
Comment (gamerson) [inline common/plugins/com.liferay.ide.ui/src/com/liferay/ide/ui/workspace/WorkbenchPreferencesTransfer.java]: It seems to be true, see here:
https://dzone.com/articles/create-new-eclipse-workspace-w

http://stackoverflow.com/a/2078547/620537

So i'm good with this implementation.

On Mon, Feb 22, 2016 at 7:16 PM, Andy Wu <notifications@github.com> wrote:

> In
> common/plugins/com.liferay.ide.ui/src/com/liferay/ide/ui/workspace/WorkbenchPreferencesTransfer.java
> <https://github.com/gamerson/liferay-ide/pull/1382#discussion_r53723061>:
>
> > +@SuppressWarnings( "restriction" )
> > +public class WorkbenchPreferencesTransfer extends WorkbenchSettingsTransfer
> > +{
> > +
> > +    public WorkbenchPreferencesTransfer()
> > +    {
> > +        super();
> > +    }
> > +
> > +    @Override
> > +    public IStatus transferSettings( IPath newWorkspaceRoot )
> > +    {
> > +        IPath currentWorkspace = Platform.getLocation();
> > +
> > +        File srcDir = new File( currentWorkspace.toFile(), ".metadata/.plugins/org.eclipse.core.runtime/.settings" );
> > +        File destDir = new File( newWorkspaceRoot.toFile(), ".metadata/.plugins/org.eclipse.core.runtime/.settings" );
>
> I tested it someway and worked well but not very sure , I read the
> http://www.vogella.com/tutorials/EclipsePreferences/article.html
> and think this is almost the right way to do .
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gamerson/liferay-ide/pull/1382/files#r53723061>.
>



-- 
Greg Amerson
Liferay Developer Tools
Liferay, Inc. www.liferay.com

File: common/plugins/com.liferay.ide.ui/src/com/liferay/ide/ui/dialog/ChooseWorkspaceWithPreferenceDialog.java (hunk 2 above)
Comment (AndyWu2015) [inline common/plugins/com.liferay.ide.ui/src/com/liferay/ide/ui/dialog/ChooseWorkspaceWithPreferenceDialog.java]: I thougnt this before. But only okPressed method in ChooseWorkspaceWithSettingsDialog  is reuseable but there is a bug in it (I mentioned in the email)  so I think It is better to keep it inheriting from ChooseWorkspaceDialog.

Comment (gamerson) [inline common/plugins/com.liferay.ide.ui/src/com/liferay/ide/ui/dialog/ChooseWorkspaceWithPreferenceDialog.java]: Ok sounds good.

On Mon, Feb 22, 2016 at 8:16 PM, Andy Wu <notifications@github.com> wrote:

> In
> common/plugins/com.liferay.ide.ui/src/com/liferay/ide/ui/dialog/ChooseWorkspaceWithPreferenceDialog.java
> <https://github.com/gamerson/liferay-ide/pull/1382#discussion_r53727771>:
>
> > +import org.eclipse.swt.widgets.Shell;
> > +import org.eclipse.ui.PlatformUI;
> > +import org.eclipse.ui.internal.WorkbenchPlugin;
> > +import org.eclipse.ui.internal.ide.ChooseWorkspaceData;
> > +import org.eclipse.ui.internal.ide.ChooseWorkspaceDialog;
> > +import org.eclipse.ui.internal.ide.IDEWorkbenchMessages;
> > +import org.eclipse.ui.internal.ide.IDEWorkbenchPlugin;
> > +import org.eclipse.ui.internal.ide.IIDEHelpContextIds;
> > +import org.eclipse.ui.internal.ide.StatusUtil;
> > +import org.eclipse.ui.preferences.SettingsTransfer;
> > +
> > +/**
> > + * @author Andy Wu
> > + */
> > +@SuppressWarnings( "restriction" )
> > +public class ChooseWorkspaceWithPreferenceDialog extends ChooseWorkspaceDialog
>
> I thougnt this before. But only okPressed method in
> ChooseWorkspaceWithSettingsDialog is reuseable but there is a bug in it (I
> mentioned in the email) so I think It is better to keep it inheriting from
> ChooseWorkspaceDialog.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gamerson/liferay-ide/pull/1382/files#r53727771>.
>



-- 
Greg Amerson
Liferay Developer Tools
Liferay, Inc. www.liferay.com
